test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java

```diff
@@ -236,34 +237,49 @@ public void initClient() throws IOException {
                     Map<?, ?> moduleInfo = (Map<?, ?>) module;
                     final String moduleName = moduleInfo.get("name").toString();
                     if (moduleName.startsWith("x-pack")) {
-                        hasXPack = true;
+                        availableFeatures.add(ProductFeature.XPACK);
                     }
                     if (moduleName.equals("x-pack-ilm")) {
-                        hasIlm = true;
+                        availableFeatures.add(ProductFeature.ILM);
+                        availableFeatures.add(ProductFeature.SLM);
                     }
                     if (moduleName.equals("x-pack-rollup")) {
-                        hasRollups = true;
+                        availableFeatures.add(ProductFeature.ROLLUPS);
                     }
                     if (moduleName.equals("x-pack-ccr")) {
-                        hasCcr = true;
+                        availableFeatures.add(ProductFeature.CCR);
                     }
                     if (moduleName.equals("x-pack-shutdown")) {
-                        hasShutdown = true;
+                        availableFeatures.add(ProductFeature.SHUTDOWN);
+                    }
+                    if (moduleName.startsWith("serverless-")) {
+                        serverless = true;
                     }
                 }
+                if (serverless) {
+                    availableFeatures.removeAll(
+                        List.of(
+                            ProductFeature.ILM,
+                            ProductFeature.SLM,
+                            ProductFeature.ROLLUPS,
+                            ProductFeature.CCR,
+                            ProductFeature.LEGACY_TEMPLATES
+                        )
+                    );
+                }
```

Comment:
I don't love what I had to do here.
1. Detecting serverless via the module names is ugly
2. It means this test case knows which APIs are available in serverless, which it shouldn't.

However, I couldn't find a better option that would require a huge rewrite of the test cases.
- There's no single test hierarchy for serverless tests, so it's hard to find somewhere to put in a shared "disable X, Y, Z" piece of code
- This method is static, so even if there was a hierarchy there isn't a great way to plug in a "disable features" piece of code.

Reply:
Is it really necessary to do this though? Should we just remove those module from serverless and then the generic code here will just "work"? Those features won't be available by virtue of the modules being uninstalled. I don't see a scenario in which we are running "serverless" _and_ have those modules installed.

Reply:
Oh how I would love that ... but alas...

- We can't remove legacy templates because it's not a module, it's part of server
- We can't remove SLM because we're using it to run operational backups, but there's no user facing API for it - everything is configured via file based config
- We can't remove ILM because SLM is part of ILM and because we support templates that configure ILM even if ILM itself cannot run

We might be able to remove CCR and Rollups, but I haven't tried since it wouldn't make a big difference.

Reply:
Understood. Where did we land on response codes here then? Could we try to hit some of these endpoints to see what they return instead. Probably more trouble than it's worth just to eliminate this small hack but just thinking out loud.

Reply:
They return `404` so I think we _could_ do it with endpoint detection, but I agree that it's not worth it (at least right now).
Either we'd do a preflight request (which requires finding a good detection API for each feature) or we'd just catch-and-ignore 404 when trying to do cleanups, which might be simpler, but could end up masking real failures.